### ISSUE TITLE
Unexpected width of candlestick candles

### ISSUE THREAD

**Issue by shrift** [patrykandpatrick/vico#923]
### How to reproduce

I use xStep with epoch offsets to graph various data according to its timestamp. To allow various density in the graphed data I change the xStep (product owners don't want to allow zoom, so we have buttons to tap to change the xStep). I noticed with larger xStep values the width of the rendered candlesticks increases.

This issue seems to be reproducible in the sample app, but to isolate the issue further I tested in the sample app with a CandleProvider and  Candle with a specific LineComponent body width.
I expected the width of the candle body to be constrained to the defined width, but instead it stretches much wider.
Changing the xStep seems to make the candle body render wider as well.
It is difficult to discern if this is just because the data points are being forced to render closer, so they're overlapping; that is expected and not what I'm reporting. Testing with very few data points makes it easier to visually reproduce the issue.

Here are the modifications to ComposeChart10 in the sample app to reproduce this. I have checked out commit `37daeb1f3a4ab92b72cd88a82de2dddb91792408`.
```
@Composable
private fun ComposeChart10(modelProducer: CartesianChartModelProducer, modifier: Modifier) {
  val marker = rememberMarker(showIndicator = false)
  val candle = CandlestickCartesianLayer.Candle(
    body = rememberLineComponent(
      color = Color.Magenta,
      thickness = 4.dp
    )
  )
  CartesianChartHost(
    chart =
      rememberCartesianChart(
        rememberCandlestickCartesianLayer(
          candles =
            object: CandlestickCartesianLayer.CandleProvider {
              override fun getCandle(
                entry: CandlestickCartesianLayerModel.Entry,
                extraStore: ExtraStore,
              ): CandlestickCartesianLayer.Candle {
                return candle
              }

              override fun getWidestCandle(extraStore: ExtraStore): CandlestickCartesianLayer.Candle {
                return candle
              }

            }
        ),
        startAxis = VerticalAxis.rememberStart(),
        bottomAxis =
          HorizontalAxis.rememberBottom(
            guideline = null,
            itemPlacer =
              remember {
                HorizontalAxis.ItemPlacer.aligned(spacing = 3, addExtremeLabelPadding = true)
              },
          ),
        marker = marker,
      ),
    modelProducer = modelProducer,
    modifier = modifier,
  )
}
```
This change adds the CandleProvider and a very basic Candle with a body LineComponent that is expected to be 4 dp wide.
To modify the data points I changed `defaultX` in `RandomCartesianModelGenerator` to ` public val defaultX: IntProgression = 0..1`.

### Observed behavior

With 2 data points and no changes other than those to chart10 already described the chart looks like this:
![image](https://github.com/user-attachments/assets/7b5bba95-c450-4736-9711-ee1ff0b9f71c)

Now adding `getXStep = { 3.0 }` to the chart It looks like this:
![image](https://github.com/user-attachments/assets/5d95f270-f71d-4c52-af6b-bed8c11b8bda)

### Expected behavior

I expected the candle body width to be constrained to the thickness defined by the candle returned in the `CandleProvider`.

### Vico version(s)

2.0.0-beta.1

### Android version(s)

14

### Additional information

I hope this all makes sense. I'm happy to help test any changes.

As always, thank you for your time and effort on Vico!

**Comment by Gowsky** [comment]
Hello, @shrift. This is expected. It's caused by the chart being zoomed in so that the content fills the viewport. This is a result of the default `Zoom` settings, which you can change. Is the following what you’re looking to achieve?

![image](https://github.com/user-attachments/assets/869fa0c0-07c0-466e-8287-627f5c4d337f)

The candles become even wider when you use `getXStep = { 3.0 }` because this causes them to overlap, meaning that the content becomes narrower intrinsically and has to be zoomed in even more to fill the viewport. (In column charts and candlestick charts, it generally doesn't make sense to use an _x_ step that's greater than the minimum _x_ delta, as it causes overlaps. If you need to change the `HorizontalAxis` label spacing, you should do so via e.g. `HorizontalAxis.ItemPlacer.aligned`.)

**Comment by shrift** [comment]
I'm not trying to change the label spacing by changing the xstep, but rather the density of the graphed data points; for instance to show one week's worth of data in the viewport or instead one month of data in the viewport.

That does appear like what I'm trying to achieve, I'll have a look at the zoom settings.

Thanks!

**Comment by Gowsky** [comment]
>I'm not trying to change the label spacing by changing the xstep, but rather the density of the graphed data points; for instance to show one week's worth of data in the viewport or instead one month of data in the viewport.

The `xStep` isn't responsible for this. It only changes what _x_ values are considered major. Instead, depending on a desired result, you can use `Zoom.x` or `CartesianLayerRangeProvider`, or you can add/remove parts of the data as required.

**Comment by shrift** [comment]
Thanks for the feedback. It seems I've misunderstood a few things. I'll dig into the things you mentioned.

**Comment by shrift** [comment]
TLDR: I overrode `LineComponent.drawVertical()` for my Candle to get the behavior I was initially expecting.
I don't have an issue now that needs to be investigated so long as overriding that function isn't expected to cause problems in the future.

I looked into this more. It seems at some point I made an assumption about what `zoomState = rememberVicoZoomState(zoomEnabled = false)` would do. I see there are some other defaults in `rememberVicoZoomState()` that affect how the charted data points are displayed. My assumption was that passing `zoomEnabled = false` would disable any affects zoom features might have on the chart (but I see now it only disables pinch to zoom gestures not the scaling affects that zoom has on how the chart renders).

I'm still confused about one point; I do need to set xStep for defining major x values for labels (for my real implementation these labels change to different timestamp formats like "11 am" "Wed" "2024" etc), but changing the xStep does also affect the density of the charted data points.
With Chart10 like this:
```
      getXStep = { 1.0 },
    ),
    zoomState = rememberVicoZoomState(
      zoomEnabled = false,
      initialZoom = Zoom.static(1f),
      minZoom = Zoom.static(1f),
      maxZoom = Zoom.static(1f),
    ),
```

I get this result:
![image](https://github.com/user-attachments/assets/402f98a6-2e51-4ca7-adf1-e048a9982767)

Changing only the xStep and leaving the same zoom I get:
```
      getXStep = { 3.0 },
    ),
    zoomState = rememberVicoZoomState(
      zoomEnabled = false,
      initialZoom = Zoom.static(1f),
      minZoom = Zoom.static(1f),
      maxZoom = Zoom.static(1f),
    ),
```
![image](https://github.com/user-attachments/assets/c314761f-8822-4927-a457-c80b07211c58)
It was a while ago now, but presumably seeing this is what lead me to controlling density of charted data with xStep instead of using zoom features.
I don't know what the relationship is exactly between xstep and zoom, but they do seem to both affect the density of the graphed data.

I still need to spend more time to see if I want to adjust zoom features at all, but It turns out the defaults are rather close to what I want.
The remaining issue for me then was that it seems my expectation was for candlestick charts to change the padding between Candles when zooms change.

I achieved this behavior this by overriding `LineComponent.drawVertical()` and passing a constant `1f` for its thicknessScale parameter.
```
  val candle = Candle(
    object : LineComponent(
      color = Color.Magenta.toArgb(),
      thicknessDp = 4.dp.value,
    ) {
      override fun drawVertical(
        context: DrawingContext,
        top: Float,
        bottom: Float,
        centerX: Float,
        thicknessScale: Float,
      ) {
        super.drawVertical(context, top, bottom, centerX, 1f)
      }
    },
  )
```
Now I can change my xStep and have my Candles render at the width I expected without setting zoom to a `Zoom.static(1f)` which it turns out is not the zoom behavior I want (as I mentioned, the default zoom behaviors are very close to my goal).
![image](https://github.com/user-attachments/assets/e9681725-5b5d-4cad-8651-788863dcbafc)

**Comment by Gowsky** [comment]
>  so long as overriding that function isn't expected to cause problems in the future

This is expected to remain functional. However, feel free to open a feature request for a parameter that excludes candle width from zoom, so that only the candle spacing is affected. There may be some edge cases in which your solution may misbehave, since the candle width may still be scaled in some calculations in `CandlestickCartesianLayer`.

> I do need to set xStep for defining major x values for labels (for my real implementation these labels change to different timestamp formats like "11 am" "Wed" "2024" etc), but changing the xStep does also affect the density of the charted data points.

As previously mentioned, you should use `HorizontalAxis.ItemPlacer` to achieve this, not the _x_-step. You mentioned that you weren't using the _x_-step to change the label spacing, but it appears that this is what you're trying to do.

For the sample chart in question, instead of `getXStep = { 3.0 }`, try `HorizontalAxis.ItemPlacer.aligned(spacing = 3)`. You'll obtain the same label spacing as in the screenshot (0, 9, 18...) without the unwanted change in candle density.

As per my previous comment, in column and candlestick charts, you generally shouldn't use an _x_-step greater than the minimum _x_ delta (unlike in line charts). Decreasing the _x_-step, however, is often appropriate. For example, if you have a column chart that displays daily data, but the data may be incomplete and have an intrinsic _x_-step of 2 instead of 1 at times, then you can lock the _x_-step at 1 so that the chart is always daily. Notice that zoom would not produce the same result — instead of shifting the columns further apart and making the chart daily, it would simply scale the content as a whole. This highlights the difference between the two.

> The remaining issue for me then was that it seems my expectation was for candlestick charts to change the padding between Candles when zooms change.

In most candlestick charts, both the candles and the spacing are scaled on zoom. However, as mentioned above, we're open to a feature request for this being made customizable.
